runtime/lua/vim/_comment.lua

```diff
@@ -3,6 +3,29 @@
 ---@field left string Left part of comment
```

Comment:
I think the whole approach can be made simpler by making the "get 'commentstring' from metdata" step directly inside the `get_commentstring()`. Here is the diff:

```diff
diff --git a/runtime/lua/vim/_comment.lua b/runtime/lua/vim/_comment.lua
index ef19793e04..db2fcea0ae 100644
--- a/runtime/lua/vim/_comment.lua
+++ b/runtime/lua/vim/_comment.lua
@@ -3,29 +3,6 @@
 ---@field left string Left part of comment
 ---@field right string Right part of comment
 
----@param row integer
----@param col integer
----@return string?
-local function get_metadata_commentstring(row, col)
-  local bufnr = vim.api.nvim_get_current_buf()
-  local caps = vim.treesitter.get_captures_at_pos(bufnr, row, col)
-  for i = #caps, 1, -1 do
-    local metadata = caps[i].metadata
-    local cms = metadata.commentstring
-    if not cms then
-      for _, value in pairs(metadata) do
-        if value.commentstring then
-          cms = value.commentstring
-          break
-        end
-      end
-    end
-    if cms then
-      return cms
-    end
-  end
-end
-
 --- Get 'commentstring' at cursor
 ---@param ref_position integer[]
 ---@return string
@@ -42,6 +19,21 @@ local function get_commentstring(ref_position)
   local row, col = ref_position[1] - 1, ref_position[2]
   local ref_range = { row, col, row, col + 1 }
 
+  -- - Get 'commentstring' from tree-sitter capture's metadata.
+  --   Traverse backwards to prefer "more narrow" capture.
+  local caps = vim.treesitter.get_captures_at_pos(0, row, col)
+  for i = #caps, 1, -1 do
+    local metadata = caps[i].metadata
+    if metadata.commentstring then
+      return metadata.commentstring
+    end
+    for _, value in pairs(metadata) do
+      if value.commentstring then
+        return value.commentstring
+      end
+    end
+  end
+
   -- - Get 'commentstring' from the deepest LanguageTree which both contains
   --   reference range and has valid 'commentstring' (meaning it has at least
   --   one associated 'filetype' with valid 'commentstring').
@@ -57,16 +49,10 @@ local function get_commentstring(ref_position)
 
     local lang = lang_tree:lang()
     local filetypes = vim.treesitter.language.get_filetypes(lang)
-
-    local metadata_cs = get_metadata_commentstring(row, col)
-    if metadata_cs and level > res_level then
-      ts_cs = metadata_cs
-    else
-      for _, ft in ipairs(filetypes) do
-        local cur_cs = vim.filetype.get_option(ft, 'commentstring')
-        if cur_cs ~= '' and level > res_level then
-          ts_cs = cur_cs
-        end
+    for _, ft in ipairs(filetypes) do
+      local cur_cs = vim.filetype.get_option(ft, 'commentstring')
+      if cur_cs ~= '' and level > res_level then
+        ts_cs = cur_cs
       end
     end
```

It basically moves (simplified version) of the approach inside `get_commentstring()` to be done *before* trying to get `'commentstring'` as languages-filetype option. Tests pass with this approach.

Reply:
Thanks! Great change. I have added a TODO here to generalize that block to a helper function, per the comment by @clason